jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +183,306 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
 
   cg_solve = partial(
       _cg_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
+
   # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  def real_valued(x):
+    return not issubclass(x.dtype.type, np.complexfloating)
   symmetric = all(map(real_valued, tree_leaves(b)))
   x = lax.custom_linear_solve(
       A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
   info = None  # TODO(shoyer): return the real iteration count here
   return x, info
+
+
+
+
+def _safe_normalize(x, return_norm=False, thresh=None):
+  """
+  Returns the L2-normalized vector (which can be a pytree) x, and optionally
+  the computed norm. If the computed norm is less than the threshold `thresh`,
+  which by default is the machine precision of x's dtype, it will be
+  taken to be 0, and the normalized x to be the zero vector.
+  """
+  norm = jnp.sqrt(jnp.abs(_vdot_tree(x, x, assume_real=False)))
```

Comment:
I would `_vdot_tree` to always return a non-negative number. Is this not the case?

Reply:
It's to cast to real

Reply:
I think this should use `assume_real=True` then? `assume_real` appears to refer to the output rather than the input variables.

Reply:
I've never used complex numbers in practice so I'm far from familiar with the conventions surrounding how they are handled numerically. I'm the one who added the `assume_real` argument here since I wasn't quite sure what the impact of the assumptions of the original implementation were. I opted to play it safe and check with @shoyer later but I never got that far with that first PR.

All that to say that it's very possible that we could just revert `_vdot_tree` to the original one without the `assume_real` argument. Someone would just have to validate that it makes sense.

Reply:
Getting rid of assume_real makes sense to me, yeah (in normal array world, I would usually write x = np.abs(np.vdot(a, b)) if I wanted a real number).